File: algorithms/algorithms-cplusplus/leetcode/Question_0006_ZigZag_Conversion.cc
#include <string>

using namespace std;

class Solution {
 public:
  string convert(string s, int numRows) {
    if (s.empty() || numRows < 2) {
      return s;
    }
    string res;
    int lag = 2 * (numRows - 1);
    for (int i = 0; i < numRows; i++) {
      for (int j = i; j < s.size(); j += lag) {
        res.push_back(s[j]);
        if (i > 0 && i < numRows - 1) {
          int n = j + lag - 2 * i;
          if (n < s.size()) {
            res.push_back(s[n]);
          }
        }
      }
    }
    return res;
  }
};

File: algorithms/algorithms-python/leetcode/Question_0006_ZigZag_Conversion.py
#!/usr/bin/python
# coding: utf-8

class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        if numRows < 2:
            return s
        res = []
        lag = 2 * (numRows - 1)
        num = len(s)
        for i in range(numRows):
            for j in range(i, num, lag):
                res.append(s[j])
                if i > 0 and i < numRows - 1:
                    n = j + lag - 2 * i
                    if n < num:
                        res.append(s[n])
        return "".join(res)
